src/anderson.jl

```diff
@@ -0,0 +1,94 @@
+# Notations from Walker & Ni, "Anderson acceleration for fixed-point iterations", SINUM 2011
+# Attempts to accelerate the iteration xn+1 = xn + β f(x)
+
+@views function anderson_{T}(df::AbstractDifferentiableMultivariateFunction,
+                      x0::Vector{T},
+                      xtol::T,
+                      ftol::T,
+                      iterations::Integer,
+                      store_trace::Bool,
+                      show_trace::Bool,
+                      extended_trace::Bool,
+                      m::Integer,
+                      β :: Real)
+
+    f_calls = 0
+    N = length(x0)
+    xs = zeros(T,N,m+1) #ring buffer storing the iterates, from newest to oldest
+    gs = zeros(T,N,m+1) #ring buffer storing the g of the iterates, from newest to oldest
+    fx = similar(x0) # temp variable to store f!
+    xs[:,1] = x0
+    errs = zeros(iterations)
+    n = 1
+    tr = SolverTrace()
+    tracing = store_trace || show_trace || extended_trace
+    old_x = xs[:,1]
+    x_converged, f_converged, converged = false, false, false
+    mat = zeros(T, N, m) #pre-allocation
+    alphas = zeros(T, m)
+
+    for n = 1:iterations
+        # fixed-point iteration
+        df.f!(xs[:,1], fx)
+        f_calls += 1
+        gs[:,1] .= xs[:,1] .+ β.*fx
+        x_converged, f_converged, converged = assess_convergence(gs[:,1], old_x, fx, xtol, ftol)
+
+        # FIXME: How should this flag be set?
+        mayterminate = false
+        
+        if tracing
+            dt = Dict()
+            if extended_trace
+                dt["x"] = copy(xs[:,1])
+                dt["f(x)"] = copy(fx)
+            end
+            update!(tr,
+                    n,
+                    maximum(abs,fx),
+                    n > 1 ? sqeuclidean(xs[:,1],old_x) : convert(T,NaN),
+                    dt,
+                    store_trace,
+                    show_trace)            
+        end
+        
+        if converged
+            break
+        end
+
+        #update of new_x
+        m_eff = min(n-1,m)
+        new_x = copy(gs[:,1])
+        if m_eff > 0
+            mat[:, 1:m_eff] .= (gs[:,2:m_eff+1] .- xs[:,2:m_eff+1]) .- (gs[:,1] .- xs[:,1])
+            alphas[1:m_eff] .= mat[:,1:m_eff] \ (xs[:,1] .- gs[:,1])
+            for i = 1:m_eff
+                new_x .+= alphas[i].*(gs[:,i+1] .- gs[:,1])
+            end
+        end
+
+        xs = circshift(xs,(0,1)) # no in-place circshift, unfortunately...
+        gs = circshift(gs,(0,1))
+        old_x = m > 1 ? xs[:,2] : copy(xs[:,1])
+        xs[:,1] = new_x
+    end
+
+    return SolverResults("Anderson m=$m β=$β",
+                         # returning gs[:,1] would be slightly better here, but the fnorm is not guaranteed
```

Comment:
Usually when solving a nonlinear problem with anderson acceleration, you expect that the base iteration xn+1 = xn + beta f(xn) is convergent, but slowly, so that returning gs[:,1], which is xs[:,1]
 + beta xs[:,1], should be slightly better than xs[:,1]. The problem is that you don't know f() of this guy, so cannot guarantee that it satisfies the convergence criterion.